### PR TITLE
Ananyamukh6/morphology_dilation

### PR DESCRIPTION
**Pull request by ananyamukh6**
Implementing morphological operation: dilation

To discuss: Assumes that images and structuring element is "binary" that is it is a float tensor with only elements from {0.0, 1.0}

### CODE REVIEW

**Comment by ananyamukh6** [comment]
screen shot of the dilation example:
dilation_example

File: examples/dilation.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import torch
+import kornia
+import cv2
+import numpy as np
+import pdb  # todo remove
+import matplotlib.pyplot as plt
+
+
+# create an image
+img = np.zeros([1, 10, 10], dtype=float)
```
Comment (edgarriba) [inline examples/dilation.py]: use torch instead of numpy

**Comment by ducha-aiki** [comment]
Thanks for contribution! I think, first we should discuss the kornia.nondifferentiable module, which is definitely where current morphology implementation belongs to.
We may want to keep both versions: classical, non-differentiable and some "soft-morphology" with sigmoids, for example.

@edgarriba FYI

File: kornia/morphology/dilation.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        raise ValueError(
+            f"Expected only 2 unique values in the tensor, since it should be binary, but got {len(torch.unique(img))}")
+    if not ((unique_vals == 0.0) + (unique_vals == 1.0)).all():
+        raise ValueError("Expected image to contain only 1's and 0's since it should be a binary image")
+
+    # Convert structuring_element from shape [a, b] to [1, 1, a, b]
+    structuring_element = structuring_element.unsqueeze(0).unsqueeze(0)
+
+    se_shape = structuring_element.shape
+    conv1 = F.conv2d(img, structuring_element, padding=(se_shape[2] // 2, se_shape[2] // 2))
```
Comment (edgarriba) [inline kornia/morphology/dilation.py]: you can use [kornia.filter2D](https://kornia.readthedocs.io/en/latest/filters.html#kornia.filters.filter2D) for this functionality

File: kornia/morphology/dilation.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        img = img.unsqueeze(0)
+    else:
+        if(img_shape[1] != 1):
+            raise ValueError(f"Expected a single channel image, but got {img_shape[1]} channels")
+    if len(structuring_element.shape) != 2:
+        raise ValueError(
+            f"Expected structuring element tensor to be of ndim=2, but got {len(structuring_element.shape)}")
+
+    # Check if the input image is a binary containing only 0, 1
+    unique_vals = torch.unique(img)
```
Comment (edgarriba) [inline kornia/morphology/dilation.py]: I would avoid this extra step and just mention it in the docs

**Comment by ducha-aiki** [comment]
> great contribution ! Just one thing: I'm not quite sure we want to place this in a separated submodule unless you have in mind to contribute with a quite large series of morphological operators. I propose to place this function inside `kornia.filters`. I recommend to discuss a bit the api and so before going ahead in the issues you open ;D

I think, we should be careful,as dilation as here is not differentiable unlike all the rest of kornia.filters functions

**Comment by ducha-aiki** [comment]
One more thing. Isn`t dilation == max-pool2d, stride=1, but with custom kernel?

File: kornia/morphology/dilation.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            f"Expected only 2 unique values in the tensor, since it should be binary, but got {len(torch.unique(img))}")
+    if not ((unique_vals == 0.0) + (unique_vals == 1.0)).all():
+        raise ValueError("Expected image to contain only 1's and 0's since it should be a binary image")
+
+    # Convert structuring_element from shape [a, b] to [1, 1, a, b]
+    structuring_element = structuring_element.unsqueeze(0).unsqueeze(0)
+
+    se_shape = structuring_element.shape
+    conv1 = F.conv2d(img, structuring_element, padding=(se_shape[2] // 2, se_shape[2] // 2))
+    convert_to_binary = (conv1 > 0).float()
```
Comment (ducha-aiki) [inline kornia/morphology/dilation.py]: This makes things non-differentiable. While we might be OK with it, I still think, that we should then create two versions: diff and non-diff

**Comment by ananyamukh6** [comment]
@edgarriba @ducha-aiki :
Its not differentiable, which is why I added it in a separate module different from "filters". I'll keep this PR open till the kornia.nondifferentiable module is available/discussed/ready-to-use, and I can move it under there.
I was thinking if we had dilation and erosion, we could have a series of easily implemented morph ops like image-open, image-close, skeleton, morphological edge detection, etc.

@ducha-aiki : Do you have some reference/link/video for the "soft" or "differentiable" morphological ops that I could look into

@edgarriba right, I know, should have probably discussed a bit before implementing it :-D . Had some free time so just started working on it <facepalm>

File: kornia/morphology/dilation.py (hunk 4 above)
Comment (edgarriba) [reply kornia/morphology/dilation.py]: probably we could call it `soft_dilation` and `dilation`

besides, form the point of view of he library I think that this last line is something that the user can do as a post process

**Comment by edgarriba** [comment]
@ananyamukh6 in that case I'm fine with a morphology module. Could you open an issue stating what functions are you planning to contribute ? we can iterate over the api there

**Comment by ducha-aiki** [comment]
https://github.com/kornia/kornia/issues/293#issuecomment-546725336

**Comment by edgarriba** [comment]
@ananyamukh6 ping on this

**Comment by ananyamukh6** [comment]
Commented on https://github.com/kornia/kornia/issues/293

**Pull request closed without merging**